rest-api/rest-api.php

```diff
@@ -0,0 +1,499 @@
+            if ( empty( $query ) ) {
+                return wp_send_json( [ 'error' => 'Bad Request' ], 400 );
+            }
+
+            global $wpdb;
+            $table_name = $wpdb->prefix . WP_Query_Push::instance()->TABLE_NAME_QUERIES;
+            $wpdb->insert(
+                $table_name,
+                [ 'query' => $query ],
+                [ "%s" ] );
+            return wp_send_json( [ 'id' => $wpdb->insert_id ], 200 );
+        } catch ( Exception $e ) {
+            return wp_send_json( [ 'error' => 'Server Error' ], 500 );
+        }
+    }
+
+    public function handle_query( WP_REST_Request $request ) {
+        try {
+            $query = $this->parse_request_query( $request );
+            if ( empty( $query ) ) {
+                return wp_send_json( [ 'error' => 'Bad Request' ], 400 );
+            }
+            return $this->run_query_json( $query );
+        } catch ( Exception $e ) {
+            return wp_send_json( [ 'error' => 'Server Error' ], 500 );
+        }
+    }
+
+    private function get_api_keys() {
+        $api_keys = get_option( 'wpquerypush_api_keys' );
+        if ( $api_keys ) {
+            return $api_keys;
+        }
+        return false;
+    }
+
+    private function get_api_keys_hashes() {
+        $api_keys = get_option( 'wpquerypush_api_keys' );
+        $api_hash = [];
+        foreach( $api_keys as $key => $value ) {
+            $api_hash[] = $value['api_hash'];
+        }
+        return $api_hash;
+    }
+
+    public function get_api_hints() {
+        $api_keys = $this->get_api_keys();
+        $api_hints = [];
+        foreach( $api_keys as $key => $value ) {
+            $api_hints[] = $value['api_hint'];
+        }
+        return $api_hints;
+    }
+
+    private function check_api_key_validity( $api_key ) {
+        $all_api_keys_hashes = $this->get_api_keys_hashes();
+        foreach ( $all_api_keys_hashes as $saved_hash ) {
+            if ( wp_check_password( $api_key, $saved_hash ) ) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private function generate_api_key() {
+        $random_bytes = random_bytes( 32 );
+        $api_key = bin2hex( $random_bytes );
+        return $api_key;
+    }
+
+    public function create_new_api_key( WP_REST_Request $request ) {
+        $new_api_key = $this->generate_api_key();
+        $existing_api_keys = get_option( 'wpquerypush_api_keys' );
+        if ( empty( $existing_api_keys ) ) {
+            $existing_api_keys = [];
+        }
+
+        $new_api_option = array( 
+            'api_hash' => wp_hash_password( $new_api_key ),
+            'api_hint' => substr( $new_api_key, -7 )
+        );
+
+        array_push( $existing_api_keys, $new_api_option );
+        update_option( 'wpquerypush_api_keys', $existing_api_keys, false );
+        return $new_api_key;
+    }
+
+    public function delete_api_key_from_hint( WP_REST_Request $request ) {
```

Comment:
@prykon I realize that it is highly, highly improbably to have only a few API Keys and have multiple end in the same sequence of 7 characters, but it was unexpected to see this.  Typically, I'd expect some sort of GUID identifier for each key, and then delete the key using that ID. This should never been a problem in our use case, but I will just note it in the Wiki.

Reply:
@zdmc23 I can create a key that gets assigned to each api hint but there's a 1 in 3,521,614,606,208,256 chance of a collision (or roughly 1 in 3.5 quadrillion). Please let me know if you still want me to add the index.